### PR TITLE
feat: allow retries on task profiling steps

### PR DESCRIPTION
**Pull request by guilhem-barthes**
## Description

(4 first commits are refactoring to `pytest`)

Fixes FL-496

## Checklist

- [ ] [changelog](../CHANGELOG.md) was updated with notable changes
- [ ] documentation was updated

### CODE REVIEW

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: backend/api/views/task_profiling.py [hunk 1]
```diff
@@ -53,14 +55,31 @@ def perform_update(self, serializer):
         return serializer.save(**kwargs)
 
 
-class TaskProfilingStepViewSet(mixins.CreateModelMixin, GenericViewSet):
+class TaskProfilingStepViewSet(mixins.CreateModelMixin, mixins.UpdateModelMixin, GenericViewSet):
     serializer_class = ProfilingStepSerializer
     authentication_classes = api_settings.DEFAULT_AUTHENTICATION_CLASSES + [BasicAuthentication]
     permission_classes = [IsAuthenticated, IsCurrentBackendOrReadOnly]
+    lookup_field = "step"
```
Comment (SdgJlbl) [inline backend/api/views/task_profiling.py:62]: this is creating an index?

File: backend/api/views/task_profiling.py [hunk 2]
```diff
@@ -53,14 +55,31 @@ def perform_update(self, serializer):
...
+        except IntegrityError:
+            data = {"detail": f"Profiling with key {task_profile_pk} and name {request.data['step']} already exists"}
+            return Response(data, status=status.HTTP_409_CONFLICT)
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+    def get_object(self):
+        queryset = self.get_queryset()
+        filters = {**self.kwargs}
+        filters["compute_task_profile_id"] = filters.pop("task_profiling_pk")
+        obj = get_object_or_404(queryset, **filters)
```
Comment (SdgJlbl) [inline backend/api/views/task_profiling.py:82]: we are not catching the eventual 404 exception?

File: backend/api/tests/views/test_views_task_profiling.py [hunk 3]
```diff
@@ -169,6 +153,41 @@ def test_task_profiling_add_step_no_datetime_change(authenticated_backend_client
...
+@override_settings(**ORG_SETTINGS)
+@pytest.mark.django_db
+def test_already_exist_step(authenticated_backend_client, task_profiling):
+    url = reverse("api:step-list", args=[str(task_profiling.compute_task.key)])
+    content = {"step": "step name", "duration": datetime.timedelta(seconds=20)}
+    response = authenticated_backend_client.post(url, content, **EXTRA)
+    assert response.status_code == status.HTTP_201_CREATED
+
+    response = authenticated_backend_client.post(url, content, **EXTRA)
+    assert response.status_code == status.HTTP_409_CONFLICT
```
Comment (SdgJlbl) [inline backend/api/tests/views/test_views_task_profiling.py:165]: I'm not sure to understand what we want to do with the 409. Will we just not display anything? will it be caught somewhere else and reconciled later?

File: backend/substrapp/tasks/tasks_compute_task.py [hunk 4]
```diff
@@ -195,27 +196,32 @@ def compute_task(self: ComputeTask, channel_name: str, serialized_task: str, com
         raise compute_task_errors.CeleryRetryError() from exception
 
 
-@retry()
-def _create_task_profiling(channel_name: str, compute_task_key: str) -> bytes:
-    parameters = {"compute_task_key": compute_task_key}
+def _send_profiling_event(*, channel_name: str, url_create: str, url_update: str, data: dict[str, Any]) -> bytes:
```
Comment (SdgJlbl) [inline backend/substrapp/tasks/tasks_compute_task.py:199]: is this shared, or just an event internal to the backend?

**Review by SdgJlbl — COMMENTED** [review]
Thanks for the PR
I have a few questions to be sure to understand how things are unfolding now :)

File: backend/api/views/task_profiling.py (hunk 2 above)
Comment (guilhem-barthes) [reply backend/api/views/task_profiling.py:82]: No, this is handled by Django, DRF. Same behavior than `get_object` in `GenericAPIView` which is used by other viewsets.

File: backend/substrapp/tasks/tasks_compute_task.py (hunk 4 above)
Comment (guilhem-barthes) [reply backend/substrapp/tasks/tasks_compute_task.py:199]: Profiling events are only registered in the api of the same backend.

File: backend/api/tests/views/test_views_task_profiling.py (hunk 3 above)
Comment (guilhem-barthes) [reply backend/api/tests/views/test_views_task_profiling.py:165]: Cf `_send_profiling_event`

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: backend/api/tests/views/conftest.py [hunk 5]
```diff
@@ -28,7 +32,14 @@ def api_client() -> test.APIClient:
 
 @pytest.fixture
 def create_compute_task():
-    def _create_compute_task(compute_plan, n_data_sample=4):
+    def _create_compute_task(
+        compute_plan: Optional[ComputePlan] = None,
+        n_data_sample: int = 4,
```
Comment (SdgJlbl) [inline backend/api/tests/views/conftest.py:37]: I know it's a port from the previous code, but any idea why this default value? (4)

**Review by SdgJlbl — APPROVED** [review]
Thanks for the PR 🙏

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Pull request merged**